monai/networks/blocks/upsample.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+import torch
+import torch.nn as nn
+from monai.networks.layers.factories import Conv
+
+
+class UpSample(nn.Module):
+    """
+    Upsample with either kernel 1 conv + interpolation or transposed conv.
+    """
+
+    def __init__(
+        self,
+        spatial_dims: int,
+        in_channels: int,
+        out_channels: Optional[int] = None,
+        scale_factor=2,
+        with_conv: bool = False,
+        mode: str = "linear",
+        align_corners: Optional[bool] = True,
+    ):
+        """
+        Args:
+            spatial_dims: number of spatial dimensions of the input image.
+            in_channels: number of channels of the input image.
+            out_channels: number of channels of the output image. Defaults to `in_channels`.
+            scale_factor: multiplier for spatial size. Has to match input size if it is a tuple. Defaults to 2.
+            with_conv: whether to use a transposed convolution for upsampling. Defaults to False.
+            mode: the interpolation mode. Defaults to "linear", this corresponds to linear, bilinear, trilinear
+                for 1D, 2D, and 3D respectively.
+            align_corners: set the align_corners parameter of `torch.nn.Upsample`. Defaults to True.
+        """
+        super().__init__()
+        if not out_channels:
+            out_channels = in_channels
+        if not with_conv:
+            _mode = ["linear", "bilinear", "trilinear"][spatial_dims - 1] if mode.lower().endswith("linear") else mode
```

Comment:
May I know why you do this mode selection logic instead of just using the `mode` parameter?
Thanks.

Reply:
Yes, not all the modes are compatible with all data shapes: https://github.com/pytorch/pytorch/blob/71af538e31547e5b1bc814c9e00323a21905baf3/torch/nn/functional.py#L3155-L3170

Reply:
This makes sense but I would put `["linear", "bilinear", "trilinear"]` into a separate local variable `linear_modes` or something like that to improve readability.